commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala

```diff
@@ -68,15 +62,19 @@ class JsonStringInput(reader: JsonReader, callback: AfterElement = AfterElementN
       expected("numeric string")
   }
 
-  def inputType: InputType = value match {
+  def inputType: InputType = reader.currentValue match {
     case JsonStringInput.ListMarker => InputType.List
     case JsonStringInput.ObjectMarker => InputType.Object
     case null => InputType.Null
     case _ => InputType.Simple
   }
 
-  def readNull(): Null = if (value == null) null else expected("null")
-  def readString(): String = matchOr[String]("string")
+  def readNull(): Null = if (reader.currentValue == null) null else expected("null")
+  def readString(): String = {
+    val expected = "string"
+    if (reader.isValueNumeric) throw new ReadFailure(s"Expected $expected but got numeric string: ${reader.currentValue}")
```

Comment:
I would say that's just "number", not a "numeric string". In JSON this is just a number, the fact that we keep it as string is our implementation detail.